src/socket.rs

```diff
@@ -226,23 +245,27 @@ impl Socket {
     /// the field in the process. This can be useful for checking errors between
     /// calls.
     pub fn take_error(&self) -> io::Result<Option<io::Error>> {
-        self.inner().take_error()
+        sys::take_error(self.inner)
     }
 
     /// Moves this TCP stream into or out of nonblocking mode.
     ///
-    /// On Unix this corresponds to calling fcntl, and on Windows this
-    /// corresponds to calling ioctlsocket.
+    /// # Notes
+    ///
+    /// On Unix this corresponds to calling `fcntl` (un)setting `O_NONBLOCK`.
+    ///
+    /// On Windows this corresponds to calling `ioctlsocket` (un)setting
+    /// `FIONBIO`.
```

Comment:
I'm not sure about Windows, but on Linux this can race with other threads that modify file status flags. It could be nice to add a recommendation not to change the flag if the FD is shared with other threads.

On the other hand, as long as `socket2` doesn't expose other file status flags, there shouldn't be practical issues until someone uses the raw FD. At that point I think it's fine to assume they should know what they are doing. So maybe the warning isn't that important ^^

Reply:
It's true. We can do two things:
1) Change the signature to `&mut self`, ensuring we got unique access and eliminating the data races.
2) Document it as you said.

As higher level library, such as Mio and Tokio, already use `&self` in there signature I think we should keep it as is and go with the second option.

I do plan on adding more options, all actually, so maybe such a warning should go into the `struct Socket` documentation?

Reply:
Documentation can't hurt, but I don't think there are any one file status flags that matter for sockets. At least on Linux, the only other flags that `F_SETFL` can change are `O_APPEND`, `O_ASYNC`, `O_DIRECT`, `O_NOATIME` and `O_NONBLOCK`.

Of those, only `O_ASYNC` and `O_NONBLOCK` have any meaning for sockets. But `O_ASYNC` is very niche: it enables signal-driven I/O. Because of the global nature of signals, that tends to be the opposite of what you want in Rust.

But still, documenting the racy nature is still good.